tss-esapi/src/abstraction/pcr/bank.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    structures::{Digest, PcrSlot},
+    Error, Result, WrapperErrorKind,
+};
+use log::error;
+use std::collections::BTreeMap;
+
+/// Struct for holding PcrSlots and their
+/// corresponding values.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct PcrBank {
+    bank: BTreeMap<PcrSlot, Digest>,
+}
+
+impl PcrBank {
+    /// Function that creates PcrBank from a vector of pcr slots and
+    /// a vector of pcr digests.
+    ///
+    /// # Details
+    /// The order of pcr slots are assumed to match the order of the Digests.
+    ///
+    /// # Error
+    /// - If number of pcr slots does not match the number of pcr digests
+    ///   InconsistentParams error is returned.
+    ///
+    /// - If the vector of pcr slots contains duplicates then
+    ///   InconsistentParams error is returned.
+    pub fn create(mut pcr_slots: Vec<PcrSlot>, mut digests: Vec<Digest>) -> Result<PcrBank> {
+        if pcr_slots.len() != digests.len() {
+            error!(
+                "Number of PcrSlots does not match the number of PCR digests. ({} != {})",
+                pcr_slots.len(),
+                digests.len()
+            );
+            return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+        }
+        pcr_slots
+            .drain(..)
+            .zip(digests.drain(..))
+            .try_fold(BTreeMap::<PcrSlot, Digest>::new(), |mut data, (pcr_slot, digest)| {
+                if data.insert(pcr_slot, digest).is_none() {
+                    Ok(data)
+                } else {
+                    error!("Error trying to insert data into PcrSlot {:?} where data have already been inserted", pcr_slot);
+                    Err(Error::local_error(WrapperErrorKind::InconsistentParams))
+                }
+            })
+            .map(|bank|  PcrBank { bank })
+    }
+
+    /// Retrieves reference to a [Digest] associated with the provided [PcrSlot].
+    ///
+    /// # Details
+    /// Returns a reference to a [Digest] associated with the provided [PcrSlot]
+    /// if one exists else returns None.
+    pub fn get_digest(&self, pcr_slot: PcrSlot) -> Option<&Digest> {
+        self.bank.get(&pcr_slot)
+    }
+
+    /// Returns true if the [PcrBank] contains a digest
+    /// for the provided [PcrSlot].
+    pub fn has_digest(&self, pcr_slot: PcrSlot) -> bool {
+        self.bank.contains_key(&pcr_slot)
+    }
+
+    /// Number of digests in the [PcrBank]
+    pub fn len(&self) -> usize {
+        self.bank.len()
+    }
+
+    /// Returns true if the [PcrBank] is empty
+    pub fn is_empty(&self) -> bool {
+        self.bank.is_empty()
```

Comment:
I think it's okay here but too bad that now we've got some types that Deref to their inner types and some that don't. But Vecs are quite a lot simpler than BTreeMaps I guess.